tests/difference/core/quint_model/ccv.qnt

```diff
@@ -0,0 +1,619 @@
+            outstandingPacketsToProvider: outstandingPacketsToProvider,
+            outstandingPacketsToConsumer: outstandingPacketsToConsumer,
+            maturationTimes: maturationTimes,
+            receivedMaturations: receivedMaturations,
+            sentVSCPackets: sentVSCPackets,
+        }
+
+
+    // utility: a map assigning each chain to false, used for not advancing consumer status
+    pure val NoStatusAdvancement: Chain -> bool = chains.mapBy(
+        (chain) => 
+        false
+    )
+
+    // utility: the set of consumers currently running
+    val RunningConsumers: Set[Chain] = 
+        ConsumerChains.filter(chain => consumerStatus.get(chain) == RUNNING)
+
+    // MODEL STATE
+    // --SHARED STATE
+
+    // Stores, for each chain, the list of voting powers that corresponded to voting powers 
+    // at blocks over its entire existence.
+    // Voting powers should be ordered by recency in descending order.
+    var votingPowerHistories: Chain -> List[ValidatorSet]
+
+    // the current validator set on each chain.
+    // this will be included in the next block, but might not be final yet,
+    // e.g. there may be more modifications in the current block.
+    var runningValidatorSet: Chain -> ValidatorSet
+
+    // the current timestamp for each chain
+    var curChainTimes: Chain -> Timestamp
+
+    // stores, for each chain, its current status -
+    // unused, running, or stopped
+    var consumerStatus: Chain -> str
+
+    // --CHANNELS
+    // Stores, for each consumer chain, the list of packets that have been sent to the provider chain
+    // and have not been received yet.
+    var outstandingPacketsToProvider: Chain -> List[VSCMaturedPacket]
+
+    // Stores, for each consumer chain, the list of packets that have been sent to the consumer chain
+    // and have not been received yet.
+    var outstandingPacketsToConsumer: Chain -> List[VSCPacket]
+
+
+    // --CONSUMER STATE
+    // Stores the maturation times for VSCPackets received by consumers
+    var maturationTimes: Chain -> (VSCPacket -> Timestamp)
+
+    // --PROVIDER STATE
+    // the set of VSCMaturedPackets received by the provider chain
+    var receivedMaturations: Set[VSCMaturedPacket]
+
+    // stores which VSC Packets have been sent to compare with receivedMaturations to detect timeouts due to non-responsiveness
+    var sentVSCPackets: Chain -> Set[VSCPacket]
+
+    // stores whether, in this step, the validator set considered to be changed.
+    // this is needed because the validator set might be considered to have changed, even though
+    // it is still technically identical at our level of abstraction, e.g. a validator power change on the provider
+    // might leave the validator set the same because a delegation and undelegation cancel each other out.
+    var providerValidatorSetChangedInThisBlock: bool
+
+
+    // UTILITY FUNCTIONS & ACTIONS
+    def wasValidatorSetOnProvider(validatorSet: ValidatorSet): bool = {
```

Comment:
Splitting the model into functional and SM would make these functions pure. For example, just define a type `VotingPowerHistories` and pass it to the function. I believe this would make the logic much easier to understand and maintain.